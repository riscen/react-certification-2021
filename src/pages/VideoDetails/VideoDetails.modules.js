import styled from 'styled-components';

export const VideoDetailsView = styled.section`
  align-items: stretch;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  overflow: scroll;
`;

export const VideoPlayer = styled.div`
  position: relative;
  width: 100%;
  height: 0;
  padding-bottom: 56.25%;
`;

export const VideoDetailsContainer = styled.div`
  flex-grow: 6;
  padding: 20px 40px;
  position: relative;
`;

export const RelatedVideosContainer = styled.div`
  flex-grow: 2;
  padding: 20px 10px;
`;
