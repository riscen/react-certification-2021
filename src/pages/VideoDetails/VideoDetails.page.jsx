import React, { useRef, useState, useEffect } from 'react';
import { useParams, withRouter } from 'react-router-dom';

import AppLayout from '../../components/AppLayout';
import RelatedVideoCard from '../../components/RelatedVideoCard/RelatedVideoCard';
import useYoutubeClient from '../../utils/hooks/useYoutubeClient';
import {
  VideoDetailsView,
  VideoPlayer,
  VideoDetailsContainer,
  RelatedVideosContainer,
} from './VideoDetails.modules';

import './VideoDetails.styles.css';

function VideoDetails(props) {
  const { get } = useYoutubeClient();
  const [videoData, setVideoData] = useState({});
  const [relatedVideos, setRelatedVideos] = useState([]);
  const youtubePlayer = useRef(null);
  const sectionRef = useRef(null);
  const { videoId } = useParams();

  const getVideos = () => {
    props.history.push('/');
  };

  useEffect(() => {
    const getVideoData = async () => {
      const videoResponse = await get('/videos', {
        id: videoId,
      });
      const relatedVideosResponse = await get('/search', {
        relatedToVideoId: videoId,
      });
      if ('items' in videoResponse.data && 'snippet' in videoResponse.data.items[0]) {
        setVideoData(videoResponse.data.items[0].snippet);
      }
      if (
        'items' in relatedVideosResponse.data &&
        'snippet' in videoResponse.data.items[0]
      ) {
        setRelatedVideos(relatedVideosResponse.data.items);
      }
    };

    getVideoData();
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <section ref={sectionRef}>
      <AppLayout handleSubmit={getVideos} />
      <VideoDetailsView>
        <VideoDetailsContainer>
          <VideoPlayer id="player" ref={youtubePlayer}>
            <iframe
              src={`https://www.youtube.com/embed/${videoId}`}
              className="video"
              title="YouTube video player"
              frameBorder="0"
              allowFullScreen
            />
          </VideoPlayer>
          <div>
            <div className="video-detail-title">
              <h2>{videoData.title && videoData.title}</h2>
            </div>
            <div className="video-detail-description">
              {videoData.description &&
                videoData.description.split('\n').map((line) => <p key={line}>{line}</p>)}
            </div>
          </div>
        </VideoDetailsContainer>
        <RelatedVideosContainer>
          {relatedVideos
            .filter((video) => 'snippet' in video)
            .map((video) => {
              return (
                <RelatedVideoCard
                  videoId={video.id.videoId}
                  title={video.snippet && video.snippet.title}
                  imageUrl={video.snippet && video.snippet.thumbnails.default.url}
                  key={video.id.videoId}
                />
              );
            })}
        </RelatedVideosContainer>
      </VideoDetailsView>
    </section>
  );
}

export default withRouter(VideoDetails);
