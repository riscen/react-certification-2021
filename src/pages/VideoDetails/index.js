export { default } from './VideoDetails.page';
export {
  VideoDetailsView,
  VideoPlayer,
  VideoDetailsContainer,
  RelatedVideosContainer,
} from './VideoDetails.modules';
