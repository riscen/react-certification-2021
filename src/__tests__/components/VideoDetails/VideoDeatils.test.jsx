import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import Adapter from 'enzyme-adapter-react-16';
import { mount, configure } from 'enzyme';

import VideoDetails, {
  VideoDetailsView,
  VideoPlayer,
  VideoDetailsContainer,
  RelatedVideosContainer,
} from '../../../pages/VideoDetails';

describe('VideoDetails', () => {
  beforeEach(() => {
    configure({ adapter: new Adapter() });
  });

  it('Renders VideoDetails Page', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('Contains VideoDetailsView', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper.find(VideoDetailsView)).toBeDefined();
    wrapper.unmount();
  });

  it('Contains VideoDetailsContainer', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper.find(VideoDetailsContainer)).toBeDefined();
    wrapper.unmount();
  });

  it('Contains Video Player', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper.find(VideoPlayer)).toBeDefined();
    expect(wrapper.find('#player')).toBeDefined();
    expect(wrapper.find('iframe')).toBeDefined();
    wrapper.unmount();
  });

  it('Displays Video Title', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper.find('.video-detail-title')).toBeDefined();
    expect(wrapper.find('h2')).toBeDefined();
    wrapper.unmount();
  });

  it('Displays Video Description', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper.find('.video-detail-description')).toBeDefined();
    wrapper.unmount();
  });

  it('Contains RelatedVideosContainer', () => {
    const wrapper = mount(
      <BrowserRouter>
        <VideoDetails />
      </BrowserRouter>
    );
    expect(wrapper.find(RelatedVideosContainer)).toBeDefined();
    wrapper.unmount();
  });
});
