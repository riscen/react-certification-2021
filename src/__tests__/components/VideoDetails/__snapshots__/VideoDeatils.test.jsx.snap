// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetails Renders VideoDetails Page 1`] = `ReactWrapper {}`;
